### PR TITLE
work in progress 1.0

### PR DESCRIPTION
**Pull request by jhcastanod**
Work in progress

### CODE REVIEW

File: src/app/interceptors/http/bad-request.interceptor.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+      catchError((error: HttpErrorResponse) => {
+        if (error.status === 400) {
+          this.formErrorsTransformer(error.error);
+        }
+        return throwError(error);
+      }
+      ));
+  }
+
+  formErrorsTransformer(error: { key: string, reason: string }[]) {
```
Comment (luqezman) [inline src/app/interceptors/http/bad-request.interceptor.ts]: switch method name to `transformErrorsToFormErrors`

File: src/app/interceptors/http/bad-request.interceptor.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import { Observable, throwError } from 'rxjs';
+import { catchError, filter } from 'rxjs/operators';
+
+export class BadRequestHttpInterceptor implements HttpInterceptor {
+  intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
+    return next.handle(req).pipe(
+      catchError((error: HttpErrorResponse) => {
+        if (error.status === 400) {
+          this.formErrorsTransformer(error.error);
+        }
```
Comment (luqezman) [inline src/app/interceptors/http/bad-request.interceptor.ts]: Add blank line here

File: src/app/interceptors/http/bad-request.interceptor.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import { HttpErrorResponse, HttpEvent, HttpHandler, HttpInterceptor, HttpRequest } from '@angular/common/http';
+
+import { Observable, throwError } from 'rxjs';
+import { catchError, filter } from 'rxjs/operators';
+
+export class BadRequestHttpInterceptor implements HttpInterceptor {
+  intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
```
Comment (luqezman) [inline src/app/interceptors/http/bad-request.interceptor.ts]: Type HttpRequest

File: src/app/interceptors/http/bad-request.interceptor.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import { HttpErrorResponse, HttpEvent, HttpHandler, HttpInterceptor, HttpRequest } from '@angular/common/http';
+
```
Comment (luqezman) [inline src/app/interceptors/http/bad-request.interceptor.ts]: Remove this line

File: src/app/interceptors/http/bad-request.interceptor.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        return throwError(error);
+      }
+      ));
+  }
+
+  formErrorsTransformer(error: { key: string, reason: string }[]) {
+    const obj = {};
+
+    for (const value of error) {
+      obj[value.key] = { IS_EMPTY: true };
```
Comment (luqezman) [inline src/app/interceptors/http/bad-request.interceptor.ts]: `obj[value.key] = { [value.reason]: true };`

File: src/app/interceptors/http/bad-request.interceptor.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import { HttpErrorResponse, HttpEvent, HttpHandler, HttpInterceptor, HttpRequest } from '@angular/common/http';
+
+import { Observable, throwError } from 'rxjs';
+import { catchError, filter } from 'rxjs/operators';
+
+export class BadRequestHttpInterceptor implements HttpInterceptor {
```
Comment (luqezman) [inline src/app/interceptors/http/bad-request.interceptor.ts]: Sort methods alphabetically

File: src/app/interceptors/http/bad-request.interceptor.ts (hunk 1 above)
Comment (luqezman) [inline src/app/interceptors/http/bad-request.interceptor.ts]: this method should be private

File: src/app/builders/heroBuilder.builder.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
+class HeroBuilder {
+  private fly = false;
+  private life = 100;
+  private name: string;
+
+  constructor(name: string) {
+    this.name = name ;
+  }
+
+  setIsFly(flying: boolean) {
```
Comment (luqezman) [inline src/app/builders/heroBuilder.builder.ts]:
```
setAsFly() {
  this.fly = true;

  return this;
}
```

File: src/app/builders/heroBuilder.builder.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
+class HeroBuilder {
+  private fly = false;
+  private life = 100;
+  private name: string;
+
+  constructor(name: string) {
+    this.name = name ;
```
Comment (luqezman) [inline src/app/builders/heroBuilder.builder.ts]: `this.name = name;`

File: src/app/builders/heroBuilder.builder.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    return this;
+  }
+
+  setLife(life: number) {
+    this.life = life;
+
+    return this;
+  }
+
+  build() {
```
Comment (luqezman) [inline src/app/builders/heroBuilder.builder.ts]: add return type `HeroModel`

File: src/app/interceptors/http/bad-request.interceptor.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
+import { HttpErrorResponse, HttpEvent, HttpHandler, HttpInterceptor, HttpRequest } from '@angular/common/http';
+
+import { Observable, throwError } from 'rxjs';
+import { catchError, filter } from 'rxjs/operators';
```
Comment (luqezman) [inline src/app/interceptors/http/bad-request.interceptor.ts]: remove filter operator

File: src/app/interceptors/http/bad-request.interceptor.ts [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+export class BadRequestHttpInterceptor implements HttpInterceptor {
+  intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
+    return next.handle(req).pipe(
+      catchError((error: HttpErrorResponse) => {
```
Comment (luqezman) [inline src/app/interceptors/http/bad-request.interceptor.ts]: add this arrow function to a private method `onCatchError`

File: src/app/interceptors/http/bad-request.interceptor.ts (hunk 1 above)
Comment (luqezman) [reply src/app/interceptors/http/bad-request.interceptor.ts]: change error to `errors`

File: src/app/interceptors/http/bad-request.interceptor.ts (hunk 3 above)
Comment (jhcastanod) [reply src/app/interceptors/http/bad-request.interceptor.ts]: what do you mean?

File: src/app/interceptors/http/bad-request.interceptor.ts (hunk 4 above)
Comment (jhcastanod) [reply src/app/interceptors/http/bad-request.interceptor.ts]: why? I don't understand this. I use them during the code

File: src/app/interceptors/http/bad-request.interceptor.ts (hunk 3 above)
Comment (luqezman) [reply src/app/interceptors/http/bad-request.interceptor.ts]: `HttpRequest<any>`

File: src/app/interceptors/http/bad-request.interceptor.ts (hunk 4 above)
Comment (luqezman) [reply src/app/interceptors/http/bad-request.interceptor.ts]: line 2

File: src/app/interceptors/http/bad-request.interceptor.ts (hunk 6 above)
Comment (jhcastanod) [reply src/app/interceptors/http/bad-request.interceptor.ts]: import { catchError, filter } from 'rxjs/operators'; **first**?
import { Observable, throwError } from 'rxjs';

how do you handle when you have **rxjs/operators** and **rxjs**?
What is the benefit of being very specific in its importation?

Comment (luqezman) [reply src/app/interceptors/http/bad-request.interceptor.ts]: import { Observable, throwError } from 'rxjs'
import { catchError, filter } from 'rxjs/operators';

File: src/app/builders/hero.builder.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+export class HeroBuilder {
+  private fly = false;
+  private life = 100;
+  private name: string;
+
+  constructor(name: string) {
+    this.name = name;
+  }
+
+  setAsFly() {
```
Comment (luqezman) [inline src/app/builders/hero.builder.ts]: add return type

File: src/app/builders/hero.builder.ts [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    this.fly = true;
+
+    return this;
+  }
+
+  setLife(life: number) {
```
Comment (luqezman) [inline src/app/builders/hero.builder.ts]: add return type

File: src/app/interceptors/bad-request.interceptor.ts [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+  HttpHandler,
+  HttpInterceptor,
+  HttpRequest,
+} from '@angular/common/http';
+import { BAD_REQUEST } from 'http-status-codes';
+import { Observable, throwError } from 'rxjs';
+import { catchError } from 'rxjs/operators';
+import { BadResquestErrorsToFormErrors } from 'transformers/bad-request-errors-to-form-errors.transformer';
+
+
```
Comment (luqezman) [inline src/app/interceptors/bad-request.interceptor.ts]: Remove this line

File: src/app/interceptors/bad-request.interceptor.ts [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+export class BadRequestHttpInterceptor implements HttpInterceptor {
+  intercept(
+    req: HttpRequest<any>,
```
Comment (luqezman) [inline src/app/interceptors/bad-request.interceptor.ts]: any is not accepted, but i Accept a todo annotation to fix it in near future

File: src/app/interceptors/bad-request.interceptor.ts [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+    next: HttpHandler,
+  ): Observable<HttpEvent<any>> {
+    return next.handle(req).pipe(catchError(this.onCatchError.bind(this)));
+  }
+
+  private onCatchError(httpError: HttpErrorResponse) {
```
Comment (luqezman) [inline src/app/interceptors/bad-request.interceptor.ts]: add return type

File: src/app/services/hero/hero.service.spec.ts [hunk 17]
```diff
@@ -1,12 +1,16 @@
 import { TestBed } from '@angular/core/testing';
```
Comment (luqezman) [inline src/app/services/hero/hero.service.spec.ts]: Fix all imports

File: src/app/transformers/bad-request-errors-to-form-errors.transformer.spec.ts [hunk 18]
```diff
@@ -0,0 +1,21 @@
+import { BadResquestErrorsToFormErrors } from './bad-request-errors-to-form-errors.transformer';
+
+describe('BadResquestErrorsToFormErrors', () => {
+  it('should transform from array to object errors structure', () => {
+    const errorFromServer = [
```
Comment (luqezman) [inline src/app/transformers/bad-request-errors-to-form-errors.transformer.spec.ts]: errors *

File: src/app/transformers/bad-request-errors-to-form-errors.transformer.spec.ts [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+      { key: 'key1', reason: 'reason1' },
+      { key: 'key2', reason: 'reason2' },
+      { key: 'key3', reason: 'reason3' },
+    ];
+    const errorExpected = {
```
Comment (luqezman) [inline src/app/transformers/bad-request-errors-to-form-errors.transformer.spec.ts]: expectedErrors *

File: src/app/transformers/bad-request-errors-to-form-errors.transformer.ts [hunk 20]
```diff
@@ -0,0 +1,13 @@
+export abstract class BadResquestErrorsToFormErrors {
+  static transformErrorsToFormErrors(errors: { key: string; reason: string; }[]) {
```
Comment (luqezman) [inline src/app/transformers/bad-request-errors-to-form-errors.transformer.ts]: add return type

File: src/app/services/hero/hero.service.ts [hunk 21]
```diff
@@ -1,30 +1,17 @@
...
-    const hero = this.createBaseHero({ fly, life, name });
-
-    return hero;
-  }
-
-  private createBaseHero(hero: IHero): IHero {
-    const mergedHero = { ...this.defaultHeroValues, ...hero };
-
-    return mergedHero;
+  create(hero: IHero) {
```
Comment (luqezman) [inline src/app/services/hero/hero.service.ts]: add return type

File: src/app/services/hero/hero.service.ts [hunk 22]
```diff
@@ -1,30 +1,17 @@
...
 @Injectable({
   providedIn: 'root'
 })
 export class HeroService {
-  readonly defaultHeroValues: IHero = { fly: false, life: 100, name: 'Aldo Raine' };
+  readonly endpoint = 'http://www.mocky.io/v2/5e7657ac2f0000710098603e';
 
-  create({ fly, life, name }: IHero): IHero {
-    const hero = this.createBaseHero({ fly, life, name });
+  constructor(private readonly http: HttpClient) { }
```
Comment (luqezman) [inline src/app/services/hero/hero.service.ts]: `{}`

File: src/app/services/hero/hero.service.ts [hunk 23]
```diff
@@ -1,30 +1,17 @@
 import { Injectable } from '@angular/core';
```
Comment (luqezman) [inline src/app/services/hero/hero.service.ts]: Sort import paths

File: src/app/services/hero/hero.service.ts (hunk 21 above)
Comment (jhcastanod) [reply src/app/services/hero/hero.service.ts]: function(): Observable<??> {}
IHero throws error
Response throws error.

What type do I expect to have here?

**Pull request merged**